### ISSUE TITLE
stack pointer is broken

### ISSUE THREAD

**Issue by ghost** [llvm-dcpu16/llvm-dcpu16#153]
Simple example

```
int div2(int a) {
  return a / 2;
}
```

```

:div2
        SUB     SP, 0x4
        SET     [0x3+SP], A
        SET     B, A
        DVI     B, 0x2
        SET     [SP], A
        SET     A, B
        ADD     SP, 0x4
        SET     PC, POP
```

As you can seem, he actually request too much stack space (I think he only needs 0x2) and uses the displacements wrong (with gaps). The same thing happens with the frame pointer of course. I think it should more look like the following:

```

:div2
        SUB     SP, 0x2
        SET     [0x1+SP], A
        SET     B, A
        DVI     B, 0x2
        SET     [SP], A
        SET     A, B
        ADD     SP, 0x2
        SET     PC, POP
```

**Comment by ghost** [comment]
Okay, I'm pretty sure that this bug originates from our change from byte to word based addressing.

**Comment by ghost** [comment]
Yep, indeed: When I change the size of a `char` to 8 again, I get the correct code (attention: code is byte based!)

```
:div2
        SUB     SP, 0x4
        SET     [0x2+SP], A
        SET     B, A
        DVI     B, 0x2
        SET     [SP], A
        SET     A, B
        ADD     SP, 0x4
        SET     PC, POP

        .globl  main
        ; .align        1
:main
        SET     PUSH, J
        SET     J, SP
        SET     [J], 0x0
        SET     A, 0x4
        JSR     div2
        SET     J, POP
        SET     PC, POP

```

**Comment by ghost** [comment]
Which is even more weird: If you use the LLVM IR with `llc -O1` he will calculate the stackpointer right -> He even uses word addressing!

**Comment by ghost** [comment]
Okay, I don't know how I should fix this problem. I tried to replace all /8 *8 hard encoded stuff with 16 bit chars, but that didn't produced the desired ASM. If anybody want's to look further into this problem: Be my guest!

**Comment by erisdev** [comment]
Commenting because this issue bit me in the ass and I'd like to know when there's progress or more information. I think this project shows the most promise out of all the compilers out there.

Sorry for not bringing anything to the discussion. C:

**Comment by Blei** [comment]
This and most of the currently open bugs are a result of 16bit bytes. Most of llvm/clang is built on the assumption that a byte is 8 bits wide (and as a result that a char is 8 bits wide as well). We are trying on a step by step basis to "fix" this, but it's a very tricky and annoying and finicky process :S

**Comment by ghost** [comment]
And you need to have a descent insight to LLVM/clang to do so. The entry step to constribute to a llvm/clang project is very high. And all developers have a rather full timetable right now. (As a sidenode: I allready spent 4-6 hours on debugging this issue)

**Comment by erisdev** [comment]
Oh dear, I hope I didn't come off like some impatient luser. I certainly don't expect this to be a quick fix, especially in a project as big as LLVM where an assumption like that is spread throughout the codebase.

Don't mind me, just overexplaining myself as usual. :heart:

Thanks for clarifying what's going on, though! Keep up the good work, guys.

**Comment by Blei** [comment]
No worries, I just wanted to clarify that there actually are people looking at open bugs but that we are not necessarily very quick in fixing them :P

**Comment by CJKay** [comment]
Generating LLVM IR with `clang -emit-llvm` and translating with the llc tool generates correct asm.
The exact same asm can be achieved by using the -fomit-frame-pointer flag inside Clang. Why that is, I don't know, but there's a temporary fix.

At least, it seems to work with one of the examples provided in one of these duplicate issues. I can't vouch for every one, of course. But feel free to write up some testcases.

**Comment by Blei** [comment]
Fixed by #167
